metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/sync/ViewSynchronizer.scala

```diff
@@ -0,0 +1,151 @@
+    *
+    * Try again until in one round we can gather all statuses from everyone.
+    */
+  def sync: F[Status[A]] = {
+    federation.publicKeys.toVector
+      .parTraverse(getAndValidateStatus)
+      .flatMap { maybeStatuses =>
+        tracers
+          .statusPoll(federation.publicKeys -> maybeStatuses)
+          .as(maybeStatuses.flatten)
+      }
+      .map(NonEmptySeq.fromSeq)
+      .flatMap {
+        case Some(statuses) if statuses.size >= federation.quorumSize =>
+          aggregateStatus(statuses).pure[F]
+
+        case _ =>
+          // We traced all responses, so we can detect if we're in an endless loop.
+          Timer[F].sleep(retryTimeout) >> sync
+      }
+  }
+
+  private def getAndValidateStatus(
+      from: A#PKey
+  ): F[Option[Validated[Status[A]]]] =
+    getStatus(from).flatMap {
+      case None =>
+        none.pure[F]
+
+      case Some(status) =>
+        validate(from, status) match {
+          case Left((error, hint)) =>
+            tracers.invalidStatus(status, error, hint).as(none)
+          case Right(valid) =>
+            valid.some.pure[F]
+        }
+    }
+
+  private def validate(
+      from: A#PKey,
+      status: Status[A]
+  ): Either[
+    (ProtocolError.InvalidQuorumCertificate[A], ViewSynchronizer.Hint),
+    Validated[Status[A]]
+  ] =
+    for {
+      _ <- validateQC(from, status.prepareQC)(
+        checkPhase(Phase.Prepare),
+        checkSignature,
+        checkVisible(status),
+        qc =>
+          check(
+            qc.viewNumber >= status.commitQC.viewNumber,
+            "Prepare Q.C. lower than Commit Q.C."
+          )
+      )
+      _ <- validateQC(from, status.commitQC)(
+        checkPhase(Phase.Commit),
+        checkSignature,
+        checkVisible(status)
+      )
+    } yield Validated[Status[A]](status)
+
+  private def check(cond: Boolean, hint: => String) =
+    if (cond) none else hint.some
+
+  private def checkPhase(phase: Phase)(qc: QuorumCertificate[A]) =
+    check(phase == qc.phase, s"Phase should be $phase.")
+
+  private def checkSignature(qc: QuorumCertificate[A]) =
+    check(Signing[A].validate(federation, qc), "Invalid signature.")
+
+  private def checkVisible(status: Status[A])(qc: QuorumCertificate[A]) =
+    check(
+      status.viewNumber >= qc.viewNumber,
+      "View number of status earlier than Q.C."
+    )
+
+  private def validateQC(from: A#PKey, qc: QuorumCertificate[A])(
+      checks: (QuorumCertificate[A] => Option[String])*
+  ) =
+    checks.toList.traverse { check =>
+      check(qc)
+        .map { hint =>
+          ProtocolError.InvalidQuorumCertificate(from, qc) -> hint
+        }
+        .toLeft(())
+    }
+}
+
+object ViewSynchronizer {
+
+  /** Extra textual description for errors. */
+  type Hint = String
+
+  /** Send a network request to get the status of a replica. */
+  type GetStatus[F[_], A <: Agreement] = A#PKey => F[Option[Status[A]]]
+
+  def aggregateStatus[A <: Agreement](
```

Comment:
This could use a similar explanation as in the PR description

Reply:
You are right, I'll add it, thanks.

Reply:
Copied descriptions and added comments to median as well.